### PR TITLE
Improve usable folder structure for downloads

### LINKED ISSUE #383: Bulking downloading folder structure
I've noticed that in practice, it's really helpful if each message/document is in it's own folder. The reason is because in Tails, you double-click a .gpg file to decrypt it, and it creates the new file in the same folder. If it's a doc, then you right-click on the .zip to extract it to get the original file that was uploaded. Sometimes (if the source is savvy and follows the journalist key instructions) this is another .gpg file with a different name, and you double-click on that to decrypt it. And sometimes, it decrypts into a .zip or .tar.gz or something that the source sent, so you extract that. Etc, etc.

Basically, each message generates at least 2 files, sometimes 3 (if they post a gpg-encrypted message in the box). And each doc generates at least 3 files (.zip.gpg, .zip, and whatever is extracted from the .zip), sometimes way more. It gets really disorganized to have all of the files just sitting there in the same folder.

(Also, sometimes a source tries to upload the same document twice because they weren't sure if it worked the first time. If the two .zip.gpg files are in the same folder, you end up extracting the zip files over each other when the contents have the same filename.)

So anyway, I propose that the bulk download feature (and any future bulk download features, such as #372) shouldn't just put all the messages/docs into a zip file, but rather should build a directory structure like this:

```
| source_codename
|--| year-month-date
|--|--| doc_number
|--|--|--| document.gpg
```

So for example, if the source codename is "Ottoman mate" and they uploaded a message and then a document on May 22, the directory structure would be:

```
ottoman-mate
ottoman-mate/2014-05-22
ottoman-mate/2014-05-22/1
ottoman-mate/2014-05-22/1/1-ottoman_mate-msg.gpg
ottoman-mate/2014-05-22/2
ottoman-mate/2014-05-22/2/2-ottoman_mate-doc.zip.gpg
```

Make sense? I don't think the directory structure needs to be exactly like this, as long as it's organized sequentially and each message/doc is in it's own folder. (And the dates are super useful, especially if there's weeks of back-and-forth between journalist and source, to look back -- but the date could just be the date of last activity.)

### PR DESCRIPTION
**Pull request by redshiftzero**
This PR is to get in the work of @ajvb from the Aaron Swartz hackathon improving the usable folder structure for downloads. Supersedes #1460 and closes #383.

### CODE REVIEW

**Comment by redshiftzero** [comment]
Rebased on current develop. @fowlslegs take a peek at this one when you get a chance

**Comment by redshiftzero** [comment]
Rebased 😇

File: securedrop/store.py [hunk 1]
```diff
@@ -68,17 +68,26 @@ def path(*s):
...
+        for source in sources:
+            submissions = [s for s in items_selected if s.source.journalist_designation == source]
+            for index, submission in enumerate(submissions, start=1):
+                filename = path(submission.source.filesystem_id,
+                                submission.filename)
+                verify(filename)
+                zip.write(filename, arcname=os.path.join(
+                    zip_directory,
+                    source,
+                    "%s_%s" % (index, submission.source.last_updated.date()),
```
Comment (fowlslegs) [inline securedrop/store.py]: This is not the index we want here. We want `submission.filename.split('_')[0]`.

File: securedrop/journalist.py [hunk 2]
```diff
@@ -665,22 +665,19 @@ def download(zip_basename, submissions):
...
-    :param list submissions: A list of :class:`db.Submission`s to 
+    :param list submissions: A list of :class:`db.Submission`s to
                              include in the zipfile.
     """
+    zf = store.get_bulk_archive(submissions,
+                                zip_directory=zip_basename)
+    attachment_filename = "{}--{}.zip".format(
+        zip_basename, datetime.utcnow().strftime("%Y-%m-%d--%H-%M-%S"))
+
     # Mark the submissions that are about to be downloaded as such
```
Comment (fowlslegs) [inline securedrop/journalist.py]: If you change the order of operations in this function, this comment uses the right tense.

**Comment by redshiftzero** [comment]
(I'm waiting on merge of #1474 to make changes here)

**Comment by fowlslegs** [comment]
You should just go ahead with this one.

**Comment by redshiftzero** [comment]
ACK

File: securedrop/store.py (hunk 1 above)
Comment (redshiftzero) [inline securedrop/store.py]: It looks to me like `index` is the right variable in order to construct the filename directory as described in [this comment](https://github.com/freedomofpress/securedrop/issues/383#issuecomment-258656017).

File: securedrop/store.py [hunk 3]
```diff
@@ -68,17 +68,26 @@ def path(*s):
     return absolute
 
 
-def get_bulk_archive(filenames, zip_directory=''):
+def get_bulk_archive(items_selected, zip_directory=''):
```
Comment (redshiftzero) [inline securedrop/store.py]: Renamed to `selected_submissions` and docstring added

File: securedrop/journalist.py (hunk 2 above)
Comment (redshiftzero) [inline securedrop/journalist.py]: Comment changed!

**Comment by redshiftzero** [comment]
OK fixed up based on your review @fowlslegs (thanks for the comments!), tests passing, and rebased on current develop. Should be good to go!

**Comment by fowlslegs** [comment]
Needs rebase.

**Comment by redshiftzero** [comment]
OK, rebased again.

**Comment by redshiftzero** [comment]
OK removed `enumerate` everywhere and rebased on current develop, this one should be good to go!

**Pull request merged**